tools/xdp/s2n-quic-xdp/src/if_xdp.rs

```diff
@@ -0,0 +1,265 @@
+#[derive(Clone, Copy, Debug, Default)]
+#[repr(C)]
+pub struct MmapOffsets<O = RingOffsetV2> {
+    pub rx: O,
+    pub tx: O,
+    pub fill: O,
+    pub completion: O,
+}
+
+impl MmapOffsets {
+    /// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L92
+    pub const RX_RING: usize = 0;
+    /// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L93
+    pub const TX_RING: usize = 0x8000_0000;
+    /// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L94
+    pub const FILL_RING: usize = 0x1_0000_0000;
+    /// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L95
+    pub const COMPLETION_RING: usize = 0x1_8000_0000;
+
+    #[inline]
+    pub(crate) fn as_v1(mut self) -> Self {
+        // getsockopt on a kernel <= 5.3 has no flags fields.
+        // Copy over the offsets to the correct places in the >=5.4 format
+        // and put the flags where they would have been on that kernel.
+
+        // Logic from https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L197
+        let v1 = unsafe { *(&self as *const Self as *const MmapOffsets<RingOffsetV1>) };
+        self.rx.set_v1(v1.rx);
+        self.tx.set_v1(v1.tx);
+        self.fill.set_v1(v1.fill);
+        self.completion.set_v1(v1.completion);
+
+        self
+    }
+}
+
+/// Socket option ids passed to an AF_XDP socket with the `setsockopt` syscall.
+///
+/// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L58-L65
+#[derive(Clone, Copy, Debug)]
+#[repr(i32)]
+pub enum SocketOptions {
+    MmapOffsets = 1,
+    RxRing = 2,
+    TxRing = 3,
+    UmemReg = 4,
+    UmemFillRing = 5,
+    UmemCompletionRing = 6,
+    Statistics = 7,
+    Options = 8,
+}
+
+/// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L67-L73
+#[derive(Clone, Copy, Debug)]
+#[repr(C)]
+pub struct UmemReg {
+    /// Start of packet data area
+    pub addr: u64,
+    /// Length of packet data area
+    pub len: u64,
+    pub chunk_size: u32,
+    /// The length reserved before packets
+    pub headroom: u32,
+    /// The offset of the flag value for the Umem
+    pub flags: u32,
+}
+
+/// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L75-L82
+#[derive(Clone, Copy, Debug, Default)]
+#[repr(C)]
+pub struct Statistics {
+    /// Dropped for other reasons
+    pub rx_dropped: u64,
+    /// Dropped due to invalid descriptor
+    pub rx_invalid_descriptors: u64,
+    /// Dropped due to invalid descriptor
+    pub tx_invalid_descriptors: u32,
```

Comment:
```suggestion
    pub tx_invalid_descriptors: u64,
```

Reply:
nice catch. that would have been annoying to debug 😄